Tasks/FtpUpload/ftpuploadtask.ts

```diff
@@ -0,0 +1,278 @@
+
+    // minimatch options
+    var matchOptions = { matchBase: true };
+    if (win) {
+        matchOptions["nocase"] = true;
+    }
+
+    tl.debug('Candidates found for match: ' + allFiles.length);
+    for (var i = 0; i < allFiles.length; i++) {
+        tl.debug('file: ' + allFiles[i]);
+    }
+
+    // use a set to avoid duplicates
+    var matchingFilesSet = new SortedSet();
+
+    for (var i = 0; i < filePatterns.length; i++) {
+        var normalizedPattern: string = path.join(rootFolder, path.normalize(filePatterns[i]));
+
+        tl.debug('searching for files, pattern: ' + normalizedPattern);
+
+        var matched = tl.match(allFiles, normalizedPattern, matchOptions);
+        tl.debug('Found total matches: ' + matched.length);
+        // ensure each result is only added once
+        for (var j = 0; j < matched.length; j++) {
+            var match = path.normalize(matched[j]);
+            if (matchingFilesSet.add(match)) {
+                var stats = tl.stats(match);
+                tl.debug('adding ' + (stats.isFile() ? 'file:   ' : 'folder: ') + match);
+                if (stats.isFile()) {
+                    var parent = path.normalize(path.dirname(match));
+                    if (matchingFilesSet.add(parent)) {
+                        tl.debug('adding folder: ' + parent);
+                    }
+                }
+            }
+        }
+    }
+    return matchingFilesSet.sorted();
+}
+
+var remotePath = tl.getInput('remotePath', true).trim();
+
+var filesUploaded: number = 0;
+var filesSkipped: number = 0; // already exists and overwrite mode off
+var directoriesProcessed: number = 0;
+
+var files = findFiles();
+
+var c = new Client();
+
+function checkDone(message: string): void {
+    var total: number = filesUploaded + filesSkipped + directoriesProcessed;
+    var remaining: number = files.length - total + 1; // add one for the root remotePath
+    console.log( 
+        'files uploaded: ' + filesUploaded +
+        ', files skipped: ' + filesSkipped +
+        ', directories processed: ' + directoriesProcessed +
+        ', total: ' + total + ', remaining: ' + remaining +
+        ', '+message);
+    if (remaining == 0) {
+        c.end();
+        tl.setResult(tl.TaskResult.Succeeded, 'FTP upload successful' + getFinalStatusMessage());
+    }
+}
+
+function getFinalStatusMessage() : string {
+    return '\nhost: ' + serverEndpointUrl.host +
+        '\npath: ' + remotePath +
+        '\nfiles uploaded: ' + filesUploaded +
+        '\nfiles skipped: ' + filesSkipped +
+        '\ndirectories processed: ' + directoriesProcessed;
+}
+
+function failTask(message: string) {
+    if(files) {
+        var total: number = filesUploaded + filesSkipped + directoriesProcessed;
+        var remaining: number = files.length - total;
+        message = message + getFinalStatusMessage() + '\nunprocessed files & directories: '+ remaining;
+    }
+    tl.setResult(tl.TaskResult.Failed, message);
```

Comment:
Be aware that setResult does not halt execution of the script.  This is why you need to pass the callback around and create valid async code.